### DISCUSSION TITLE
posts/csharp-duck-types/

### DISCUSSION THREAD

**Discussion by giscus** [BYJRK/byjrk.github.io#2] [Announcements]
# posts/csharp-duck-types/

这篇文章我们来总结一下 C# 中的那些不为人知的鸭子类型。

https://blog.coldwind.top/posts/csharp-duck-types/

**Comment by BYJRK** [comment]
P.S. 本文的封面是使用 DALL·E 绘制的。

<details>
<summary>Prompts</summary>
<p>A creative and engaging cover image for a blog post about duck typing in programming, blending the worlds of nature and technology. The scene includes cartoon ducks sitting in front of computer screens, typing away on keyboards with intense focus. Around them, binary code floats in the air like bubbles, symbolizing the coding aspect. In the background, there's a digital landscape, representing the virtual environment where programming occurs. This digital landscape has a matrix-like appearance, with streams of code running down the screens the ducks are working on. The ducks wear glasses and have expressions of deep concentration, emphasizing their involvement in the coding process. The image is designed with a wide aspect ratio to fit perfectly as a blog post cover.</p>
</details>

**Comment by BYJRK** [comment]
我的天，我写完之后才发现，这篇文章和 [hez2010 的文章](https://zhuanlan.zhihu.com/p/121704199) 高度撞车……

**Reply by BYJRK** [reply]
我姑且比他多了个集合初始化器😂但是他又比我多了一些诸如 IAsyncEnumerable、Range、Index 等

**Comment by CalmlyCC** [comment]
大佬，为啥说SelectMany比较冷门且不常用呀😂

**Reply by BYJRK** [reply]
不是 `SelectMany` 不常用，而是单独给一个没实现 `IEnumerable` 接口，却又实现了 `SelectMany` 方法的情况太冷门

**Comment by wangsen3** [comment]
楼主真厉害, 在哪里上班啊

**Comment by Renllyy** [comment]
点赞

**Comment by AnyThing-Well** [comment]
我知道鸭子类型还是因为有一次逛reddit的时候发现了一个有趣的扩展
public static TaskAwaiter<(T1, T2)> GetAwaiter<T1, T2>(this (Task<T1>, Task<T2>) tasks)
    => WhenAllResult(tasks).GetAwaiter();

public static async Task<(T1, T2)> WhenAllResult<T1, T2>(this (Task<T1>, Task<T2>) tasks)
{
    await Task.WhenAll(tasks.Item1, tasks.Item2).ConfigureAwait(false);
    return (tasks.Item1.Result, tasks.Item2.Result);
}

var (result1, result2) = await (
    GetDataAsync(),
    GetOtherDataAsync()
);

**Reply by BYJRK** [reply]
以后的async 2.0可能直接自带这个语法了
